### PR TITLE
[RF] Improve constraint parameter management

### PR DESCRIPTION
**Pull request by guitargeek**
The `Constrain()` and `GlobalObservables()` arguments were wrong in a
unit test in `testLikelihoodGradientJob`: a global observable was added
as a constrained parameter, which doesn't make sense.

`Constrain()` should contain the constrained parameters, and
`GlobalObservables()` the global observables.

### CODE REVIEW

**Review by egpbos — COMMENTED** [review]
About the first commit: I must admit that I don't know, I just copied this example. I asked Wouter to take a look.

The changes in the third commit look good to me. Cross-checking with BatchMode is a very helpful addition. I'm not sure if it makes 100% sense as part of the LikelihoodGradientJob test, seems like it should be a BatchMode test, but maybe I'm misunderstanding the bug that was fixed. I don't really mind either way.

Note that it would probably make more sense later to generalize this test to a parameterized test that tests the matrix of BatchMode true/false vs MultiProcess-enabled true/false (except the false, false case, which is just the old-style again). That would get rid of some code duplication. It's probably easier to do this when the MultiProcess-enabled stuff is also available through `fitTo`. @Zeff020 could maybe comment on the timeline on this.

**Comment by guitargeek** [comment]
Hi @egpbos, thanks for the review! You are right, the batch mode stuff should better be tested somewhere else, so I removed some commits from this PR.

What is left in this PR is the question about the constrain settings that might not make sense. Let's discuss this in the next RooFit meeting then!

What I ultimately want to achieve is that RooFit prints errors if you have constrained settings that statistically don't make sense, to reduce the potential for user error. That means we can't have "unphysical" fits in our tests too, otherwise they would fail eventually from these error messages.

**Comment by egpbos** [comment]
Makes total sense. If you're going to reuse the testcase for batch mode tests later on, it would probably still be good to extract the common code into a utility header. I liked a lot the cleanup you already did before on this test file.

File: roofit/roofitcore/test/TestStatistics/testLikelihoodJob.cxx [hunk 1]
```diff
@@ -109,7 +109,8 @@ TEST_F(LikelihoodJobTest, UnbinnedGaussian1D)
 {
    std::tie(nll, pdf, data, values) = generate_1D_gaussian_pdf_nll(w, 10000);
    likelihood = RooFit::TestStatistics::buildLikelihood(pdf, data);
-   auto nll_ts = LikelihoodWrapper::create(RooFit::TestStatistics::LikelihoodMode::multiprocess, likelihood, clean_flags);
+   auto nll_ts =
+      LikelihoodWrapper::create(RooFit::TestStatistics::LikelihoodMode::multiprocess, likelihood, clean_flags);
```
Comment (egpbos) [inline roofit/roofitcore/test/TestStatistics/testLikelihoodJob.cxx:113]: Perhaps it would be better to change this pure style change and similar ones below back to not bury the actual changes in the noise.

File: roofit/roofitcore/src/RooAbsPdf.cxx [hunk 2]
```diff
@@ -1027,6 +1027,12 @@ RooAbsReal* RooAbsPdf::createNLL(RooAbsData& data, const RooLinkedList& cmdList)
   pc.defineMutex("GlobalObservables","GlobalObservablesTag") ;
   pc.defineInt("NewStyle", "NewStyle", 0, 0);
 
+  if(pc.getSet("cPars") && pc.getSet("extCons")) {
+    std::string errMsg = "Specifying Constrain() when GlobalObservables() is passed is dangerous!";
+    coutE(InputArguments) << errMsg << std::endl;
+    throw std::invalid_argument(errMsg);
+  }
```
Comment (egpbos) [inline roofit/roofitcore/src/RooAbsPdf.cxx:1034]: Wouldn't it be easier to just use `pc.defineMutex` here rather than doing this manually?

File: roofit/roofitcore/test/TestStatistics/testLikelihoodJob.cxx [hunk 3]
```diff
@@ -416,8 +426,8 @@ TEST_F(LikelihoodJobSimBinnedConstrainedTest, BasicParameters)
 TEST_F(LikelihoodJobSimBinnedConstrainedTest, ConstrainedAndOffset)
 {
    // a variation to test some additional parameters (ConstrainedParameters and offsetting)
-   nll.reset(pdf->createNLL(*data, RooFit::Constrain(RooArgSet(*w.var("alpha_bkg_obs_A"))),
-                            RooFit::GlobalObservables(RooArgSet(*w.var("alpha_bkg_obs_B"))), RooFit::Offset(true)));
+   nll.reset(
+      pdf->createNLL(*data, RooFit::GlobalObservables(RooArgSet(*w.var("alpha_bkg_obs_B"))), RooFit::Offset(true)));
```
Comment (egpbos) [inline roofit/roofitcore/test/TestStatistics/testLikelihoodJob.cxx:430]: `alpha_bkg_obs_A` must also be in the global observables set, I think. This is probably what is breaking the CI test currently.

File: roofit/roofitcore/test/TestStatistics/testLikelihoodJob.cxx [hunk 4]
```diff
@@ -475,13 +487,14 @@ TEST_F(LikelihoodJobTest, BatchedUnbinnedGaussianND)
...
+class LikelihoodJobSplitStrategies : public LikelihoodJobSimBinnedConstrainedTest,
+                                     public testing::WithParamInterface<std::tuple<std::size_t, std::size_t>> {};
 
 TEST_P(LikelihoodJobSplitStrategies, SimBinnedConstrainedAndOffset)
 {
    // based on ConstrainedAndOffset, this test tests different parallelization strategies
-   nll.reset(pdf->createNLL(*data, RooFit::Constrain(RooArgSet(*w.var("alpha_bkg_obs_A"))),
-                            RooFit::GlobalObservables(RooArgSet(*w.var("alpha_bkg_obs_B"))), RooFit::Offset(true)));
+   nll.reset(
+      pdf->createNLL(*data, RooFit::GlobalObservables(RooArgSet(*w.var("alpha_bkg_obs_B"))), RooFit::Offset(true)));
```
Comment (egpbos) [inline roofit/roofitcore/test/TestStatistics/testLikelihoodJob.cxx:497]: Also here add `A`.

File: roofit/roofitcore/test/TestStatistics/testLikelihoodJob.cxx [hunk 5]
```diff
@@ -517,12 +531,11 @@ TEST_P(LikelihoodJobSplitStrategies, SimBinnedConstrainedAndOffset)
...
-    )
-);
+INSTANTIATE_TEST_SUITE_P(SplitStrategies, LikelihoodJobSplitStrategies,
+                         testing::Combine(
+                            // number of event tasks:
+                            testing::Values(0, 1, 2, 50, 100, 101, 102,
+                                            100000), // the last value is larger than number of events to test that
+                            // number of component tasks:
+                            testing::Values(0, 1, 2, 4) // the last value is larger than number of components
+                            ));
```
Comment (egpbos) [inline roofit/roofitcore/test/TestStatistics/testLikelihoodJob.cxx:541]: Revert style change.

Comment (egpbos) [reply roofit/roofitcore/test/TestStatistics/testLikelihoodJob.cxx:541]: Or do in separate commit, of course.

File: roofit/roostats/src/AsymptoticCalculator.cxx [hunk 6]
```diff
@@ -306,7 +306,7 @@ double AsymptoticCalculator::EvaluateNLL(RooAbsPdf & pdf, RooAbsData& data,   co
 
     // need to call constrain for RooSimultaneous until stripDisconnected problem fixed
     auto& config = GetGlobalRooStatsConfig();
-    RooAbsReal* nll = pdf.createNLL(data, RooFit::CloneData(false),RooFit::Constrain(*allParams),RooFit::ConditionalObservables(conditionalObs), RooFit::GlobalObservables(globalObs),
+    RooAbsReal* nll = pdf.createNLL(data, RooFit::CloneData(false),RooFit::ConditionalObservables(conditionalObs), RooFit::GlobalObservables(globalObs),
```
Comment (egpbos) [inline roofit/roostats/src/AsymptoticCalculator.cxx:309]: Could these be edge cases where `Constrain` still needs to be supported because it was (ab)used before?

**Review by egpbos — CHANGES_REQUESTED** [review]
Some suggestions, hope they are helpful.

File: roofit/roofitcore/test/TestStatistics/testLikelihoodJob.cxx (hunk 1 above)
Comment (guitargeek) [reply roofit/roofitcore/test/TestStatistics/testLikelihoodJob.cxx:113]: Yes you're right, thanks for spotting that! I haven't realized the file was not completely formatted before

File: roofit/roofitcore/src/RooAbsPdf.cxx (hunk 2 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooAbsPdf.cxx:1034]: I don't actually want to get the exception, just print an error message. This change in RooAbsPdf is just to see in the CI how often `Constrain` is used together with `GlobalObservables`, the commit will be removed once I see there are none

Comment (egpbos) [reply roofit/roofitcore/src/RooAbsPdf.cxx:1034]: Ah, good idea!

**Comment by egpbos** [comment]
Great that you're picking this up again!

**Comment by guitargeek** [comment]
* superseded by other improvement PRs to the test statistics test
* I don't plan to remove `RooFit::Constrain()` after all, because I realized some users are relying on it to constrain only a subset of nuisance parameters for checks.
